_posts/2018-02-15-totw-142.md

```diff
@@ -0,0 +1,169 @@
+---
+title: "Tip of the Week #142: Multi-parameter Constructors and `explicit`"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/142
+type: markdown
+order: "142"
+---
+
+Originally published as totw/142 on 2018-01-29
+
+*By James Dennett [(jdennett@google.com)](mailto:jdennett@google.com)*
+
+“Explicit is better than implicit.” -- [PEP
+20](https://www.python.org/dev/peps/pep-0020/)
+
+## TL;DR:
+
+Most constructors should be explicit.
+
+## Introduction
+
+Prior to C++11, the `explicit` keyword was meaningful only for constructors that
+could be called with a single argument, and our style guide required its use for
+such constructors so that they did not act as “converting constructors”. That
+requirement was not applied for multi-parameter constructors. Indeed the style
+guide used to discourage use of `explicit` for multi-parameter constructors as
+it had no meaning. That’s no longer the case.
+
+In C++11, `explicit` became meaningful for copy-initialization from braced
+lists, such as when calling a function `void f(std::pair<int, int>)` with
+`f({1, 2})` or when initializing a variable `bad` with
+`std::vector<char> bad = {“hello”, “world”};`.
+
+Wait a minute! In that last example the types don’t match. That can’t compile,
+can it? `std::vector<string> good = {“hello”, “world”};` would be reasonable,
```

Comment:
std::string (throughout)

Reply:
Done in https://github.com/abseil/abseil.github.io/pull/145